AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClient.swift

```diff
@@ -181,48 +186,69 @@ final public class AWSMobileClient: _AWSMobileClient {
                 completionHandler(self.currentUserState, nil)
                 return
             }
+            _internalInitialize(completionHandler)
+            isInitialized = true
+        }
+    }
 
+    // Internal initialize method, pass userpoolHandler for testing purposes only.
+    internal func _internalInitialize(
+        userPoolHandler: UserPoolOperationsHandler = .sharedInstance,
+        _ completionHandler: @escaping (UserState?, Error?) -> Void) {
+        do {
             keychain.migrateToCurrentAccessibility()
+            userpoolOpsHelper = userPoolHandler
             cleanupPreviousInstall()
             initializeKeychainItems()
             fallbackLegacyFederationProvider()
 
             DeviceOperations.sharedInstance.mobileClient = self
 
-            do {
-                try registerIfPresentHostedUI()
-            } catch {
-                completionHandler(nil, error)
-            }
+            try registerIfPresentHostedUI()
 
             setIfPresentCustomAuth()
             setIfPresentCredentialsProvider()
 
-            let userState = determineIntialUserState()
-            if userState == .signedIn
-                && (federationProvider == .userPools || federationProvider == .hostedUI)
-                && self.username == nil {
-                self.signOut()
-                currentUserState = .signedOut
-            } else {
-                currentUserState = userState
-            }
+            currentUserState = determineIntialUserState()
             completionHandler(currentUserState, nil)
-            isInitialized = true
+        } catch {
+            completionHandler(nil, error)
         }
     }
 
+    /// Using the cached keychain items determine the user state.
     private func determineIntialUserState() -> UserState {
+        var userState: UserState = .signedOut
         if (self.cachedLoginsMap.count > 0) {
-            return .signedIn
+            userState = .signedIn
+
+        } else if let credentialProvider = self.internalCredentialsProvider,
+                  credentialProvider.identityId != nil {
+            userState = (federationProvider == .none) ? .guest : .signedIn
         }
 
-        if let credentialProvider = self.internalCredentialsProvider,
-           credentialProvider.identityId != nil {
-            return (federationProvider == .none) ? .guest : .signedIn
+        // SignOut if we get an invalid signedIn state
+        if userState == .signedIn
+            && !isValidSignedIn (
+                userState: userState,
+                federationProvider: federationProvider
+            ) {
+            AWSMobileClientLogging.verbose("Invalid signedIn state found, signing out")
+            signOut()
+            userState = .signedOut
         }
-        return .signedOut
+        return userState
     }
+
+    private func isValidSignedIn(
```

Comment:
suggestion to rename the function : `isValidSignedInState`

Reply:
Changed